lib/Controller/MessagesController.php

```diff
@@ -710,35 +728,63 @@ public function removeTag(int $id, string $imapLabel): JSONResponse {
 		return new JSONResponse($tag);
 	}
 
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * Messages are not expected to be deleted one by one anymore
+	 * => function deleteMessages is expected to be used.
+	 * @param int $id
+	 *
+	 * @throws NotImplemented
+	 */
+	public function destroy(int $id): void {
+		throw new \OCA\Mail\Exception\NotImplemented();
+	}
+
 	/**
 	 * @NoAdminRequired
 	 * @TrapError
 	 *
 	 * @param int $accountId
 	 * @param string $folderId
-	 * @param int $id
+	 * @param int[] $ids
 	 *
 	 * @return JSONResponse
 	 *
-	 * @throws ClientException
-	 * @throws ServiceException
 	 */
-	public function destroy(int $id): JSONResponse {
-		try {
+	public function deleteMessages(array $ids): JSONResponse {
+		// Creates a deletion batch subdivised in accounts and mailboxes
+		$batch = [];
+		foreach ($ids as $id) {
 			$message = $this->mailManager->getMessage($this->currentUserId, $id);
 			$mailbox = $this->mailManager->getMailbox($this->currentUserId, $message->getMailboxId());
-			$account = $this->accountService->find($this->currentUserId, $mailbox->getAccountId());
-		} catch (DoesNotExistException $e) {
-			return new JSONResponse([], Http::STATUS_FORBIDDEN);
+			$mailboxName = $mailbox->getName();
+			$accountId = $mailbox->getAccountId();
+			if (!array_key_exists($accountId, $batch)) {
+				$batch[$accountId] = [];
+			}
+			if (!array_key_exists($mailboxName, $batch[$accountId])) {
+				$batch[$accountId][$mailboxName] = [];
+			}
+			array_push($batch[$accountId][$mailboxName],$message->getUid());
 		}
 
-		$this->logger->debug("deleting message <$id>");
+		// Deletes messages from batch
+		try {
+			foreach ($batch as $accountId => $subbatch) {
+				foreach ($subbatch as $mailboxName => $uids) {
+					$this->mailManager->deleteMessages(
```

Comment:
perf: there will be separate IMAP connections for every deletion. this will not scale. we have to reuse the horde imap client object